[GStreamer] WebAudio drums demo makes WebKit GStreamer based ports crash

philn opened the pull request:
#### 2aa13bb90ada0abbfe5ec6e6ad1fcc95c798a361
<pre>
[GStreamer] WebAudio drums demo makes WebKit GStreamer based ports crash
<a href="https://bugs.webkit.org/show_bug.cgi?id=244841">https://bugs.webkit.org/show_bug.cgi?id=244841</a>

Reviewed by Xabier Rodriguez-Calvar.

The main issue was a mix-up between channel index and channel types. The m_buffers hash table stores
buffers by channel type, so when we need to copy the data to the bus, we need to use
AudioBus::channelByType().

* Source/WebCore/platform/audio/gstreamer/AudioFileReaderGStreamer.cpp:
(WebCore::channelTypeFromCaps):
(WebCore::AudioFileReader::handleSample):
(WebCore::AudioFileReader::handleNewDeinterleavePad):
(WebCore::AudioFileReader::createBus):

Canonical link: <a href="https://commits.webkit.org/254223@main">https://commits.webkit.org/254223@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6ff696791b4fb2dfc82a09df70bbc3107da42b3f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/88450 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/32942 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/19305 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/97626 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/153104 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/92414 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/31438 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/27056 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/80645 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/92281 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/94053 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/24987 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/75339 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/24948 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/79877 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/79978 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/67911 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/29046 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/13963 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/29030 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/14985 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2974 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/32254 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/37911 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/31132 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/34094 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/platform/audio/gstreamer/AudioFileReaderGStreamer.cpp:
```diff
@@ -182,14 +171,40 @@ GstFlowReturn AudioFileReader::handleSample(GstAppSink* sink)
+    if (!channelType)
+        return GST_FLOW_ERROR;
+
+    RELEASE_ASSERT(m_firstChannelType);
```
I would make this a `GST_FLOW_ERROR` as well, maybe combined with `ASSERT_NOT_REACHED` if you want to be in the safer side or just an `ASSERT`

philn committed:
[GStreamer] WebAudio drums demo makes WebKit GStreamer based ports crash
https://bugs.webkit.org/show_bug.cgi?id=244841

Reviewed by Xabier Rodriguez-Calvar.

The main issue was a mix-up between channel index and channel types. The m_buffers hash table stores
buffers by channel type, so when we need to copy the data to the bus, we need to use
AudioBus::channelByType().

* Source/WebCore/platform/audio/gstreamer/AudioFileReaderGStreamer.cpp:
(WebCore::channelTypeFromCaps):
(WebCore::AudioFileReader::handleSample):
(WebCore::AudioFileReader::handleNewDeinterleavePad):
(WebCore::AudioFileReader::createBus):

Canonical link: https://commits.webkit.org/254223@main
